// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/core/law-cli-error/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelegateeErrors = exports.RemoveToolPolicyParameterErrors = exports.SetToolPolicyParameterErrors = exports.GetToolPolicyParameterErrors = exports.UnpermitToolForDelegateeErrors = exports.PermitToolForDelegateeErrors = exports.ManageDelegateesMenuErrors = exports.IsDelegateeErrors = exports.RemoveDelegateeErrors = exports.AddDelegateeErrors = exports.GetDelegateesErrors = exports.GetToolPolicyErrors = exports.DisablePolicyErrors = exports.EnablePolicyErrors = exports.RemovePolicyErrors = exports.SetPolicyErrors = exports.ManagePoliciesMenuErrors = exports.DisableToolErrors = exports.EnableToolErrors = exports.RemoveToolErrors = exports.PermitToolErrors = exports.AdminErrors = exports.RemoveRpcErrors = exports.EditRpcErrors = exports.AddRpcErrors = exports.ManageRpcsMenuErrors = exports.ChangeLitNetworkErrors = exports.CliSettingsMenuErrors = exports.MainMenuErrors = exports.GetLitNetworkErrors = exports.GeneralErrors = void 0;
var GeneralErrors;
(function (GeneralErrors) {
    GeneralErrors["UNKNOWN_LAW_CLI_ERROR"] = "UNKNOWN_LAW_CLI_ERROR";
    GeneralErrors["NO_ACTION_SELECTED"] = "NO_ACTION_SELECTED";
})(GeneralErrors || (exports.GeneralErrors = GeneralErrors = {}));
var GetLitNetworkErrors;
(function (GetLitNetworkErrors) {
    GetLitNetworkErrors["NO_LIT_NETWORK_SELECTED"] = "NO_LIT_NETWORK_SELECTED";
})(GetLitNetworkErrors || (exports.GetLitNetworkErrors = GetLitNetworkErrors = {}));
var MainMenuErrors;
(function (MainMenuErrors) {
    MainMenuErrors["NO_MAIN_MENU_ACTION_SELECTED"] = "NO_MAIN_MENU_ACTION_SELECTED";
    MainMenuErrors["UNKNOWN_MAIN_MENU_ACTION"] = "UNKNOWN_MAIN_MENU_ACTION";
})(MainMenuErrors || (exports.MainMenuErrors = MainMenuErrors = {}));
var CliSettingsMenuErrors;
(function (CliSettingsMenuErrors) {
    CliSettingsMenuErrors["NO_CLI_SETTINGS_ACTION_SELECTED"] = "NO_CLI_SETTINGS_ACTION_SELECTED";
    CliSettingsMenuErrors["UNKNOWN_CLI_SETTINGS_ACTION"] = "UNKNOWN_CLI_SETTINGS_ACTION";
})(CliSettingsMenuErrors || (exports.CliSettingsMenuErrors = CliSettingsMenuErrors = {}));
var ChangeLitNetworkErrors;
(function (ChangeLitNetworkErrors) {
    ChangeLitNetworkErrors["NO_LIT_NETWORK_SELECTED"] = "NO_LIT_NETWORK_SELECTED";
})(ChangeLitNetworkErrors || (exports.ChangeLitNetworkErrors = ChangeLitNetworkErrors = {}));
var ManageRpcsMenuErrors;
(function (ManageRpcsMenuErrors) {
    ManageRpcsMenuErrors["UNKNOWN_RPC_ACTION"] = "UNKNOWN_RPC_ACTION";
})(ManageRpcsMenuErrors || (exports.ManageRpcsMenuErrors = ManageRpcsMenuErrors = {}));
var AddRpcErrors;
(function (AddRpcErrors) {
    AddRpcErrors["CHAIN_NAME_EXISTS"] = "CHAIN_NAME_EXISTS";
    AddRpcErrors["ADD_RPC_CANCELLED"] = "ADD_RPC_CANCELLED";
})(AddRpcErrors || (exports.AddRpcErrors = AddRpcErrors = {}));
var EditRpcErrors;
(function (EditRpcErrors) {
    EditRpcErrors["EDIT_RPC_CANCELLED"] = "EDIT_RPC_CANCELLED";
    EditRpcErrors["NO_RPCS_FOUND"] = "NO_RPCS_FOUND";
})(EditRpcErrors || (exports.EditRpcErrors = EditRpcErrors = {}));
var RemoveRpcErrors;
(function (RemoveRpcErrors) {
    RemoveRpcErrors["REMOVE_RPC_CANCELLED"] = "REMOVE_RPC_CANCELLED";
    RemoveRpcErrors["NO_RPCS_FOUND"] = "NO_RPCS_FOUND";
})(RemoveRpcErrors || (exports.RemoveRpcErrors = RemoveRpcErrors = {}));
var AdminErrors;
(function (AdminErrors) {
    AdminErrors["ADMIN_MISSING_PRIVATE_KEY"] = "ADMIN_MISSING_PRIVATE_KEY";
    AdminErrors["FAILED_TO_INITIALIZE_ADMIN"] = "FAILED_TO_INITIALIZE_ADMIN";
    AdminErrors["NO_PKPS_FOUND"] = "NO_PKPS_FOUND";
    AdminErrors["PKP_SELECTION_CANCELLED"] = "PKP_SELECTION_CANCELLED";
    AdminErrors["ADMIN_SELECTION_CANCELLED"] = "ADMIN_SELECTION_CANCELLED";
})(AdminErrors || (exports.AdminErrors = AdminErrors = {}));
var PermitToolErrors;
(function (PermitToolErrors) {
    PermitToolErrors["PERMIT_TOOL_CANCELLED"] = "PERMIT_TOOL_CANCELLED";
    PermitToolErrors["NO_UNPERMITTED_TOOLS"] = "NO_UNPERMITTED_TOOLS";
    PermitToolErrors["ENABLE_TOOL_CANCELLED"] = "ENABLE_TOOL_CANCELLED";
})(PermitToolErrors || (exports.PermitToolErrors = PermitToolErrors = {}));
var RemoveToolErrors;
(function (RemoveToolErrors) {
    RemoveToolErrors["REMOVE_TOOL_CANCELLED"] = "REMOVE_TOOL_CANCELLED";
    RemoveToolErrors["NO_PERMITTED_TOOLS"] = "NO_PERMITTED_TOOLS";
})(RemoveToolErrors || (exports.RemoveToolErrors = RemoveToolErrors = {}));
var EnableToolErrors;
(function (EnableToolErrors) {
    EnableToolErrors["ENABLE_TOOL_CANCELLED"] = "ENABLE_TOOL_CANCELLED";
    EnableToolErrors["NO_DISABLED_TOOLS"] = "NO_DISABLED_TOOLS";
})(EnableToolErrors || (exports.EnableToolErrors = EnableToolErrors = {}));
var DisableToolErrors;
(function (DisableToolErrors) {
    DisableToolErrors["DISABLE_TOOL_CANCELLED"] = "DISABLE_TOOL_CANCELLED";
    DisableToolErrors["NO_ENABLED_TOOLS"] = "NO_ENABLED_TOOLS";
})(DisableToolErrors || (exports.DisableToolErrors = DisableToolErrors = {}));
var ManagePoliciesMenuErrors;
(function (ManagePoliciesMenuErrors) {
    ManagePoliciesMenuErrors["NO_MANAGE_POLICIES_ACTION_SELECTED"] = "NO_MANAGE_POLICIES_ACTION_SELECTED";
})(ManagePoliciesMenuErrors || (exports.ManagePoliciesMenuErrors = ManagePoliciesMenuErrors = {}));
var SetPolicyErrors;
(function (SetPolicyErrors) {
    SetPolicyErrors["SET_POLICY_CANCELLED"] = "SET_POLICY_CANCELLED";
    SetPolicyErrors["NO_TOOLS_FOUND"] = "NO_TOOLS_FOUND";
})(SetPolicyErrors || (exports.SetPolicyErrors = SetPolicyErrors = {}));
var RemovePolicyErrors;
(function (RemovePolicyErrors) {
    RemovePolicyErrors["REMOVE_POLICY_CANCELLED"] = "REMOVE_POLICY_CANCELLED";
    RemovePolicyErrors["NO_POLICIES_FOUND"] = "NO_POLICIES_FOUND";
})(RemovePolicyErrors || (exports.RemovePolicyErrors = RemovePolicyErrors = {}));
var EnablePolicyErrors;
(function (EnablePolicyErrors) {
    EnablePolicyErrors["ENABLE_POLICY_CANCELLED"] = "ENABLE_POLICY_CANCELLED";
    EnablePolicyErrors["NO_DISABLED_POLICIES"] = "NO_DISABLED_POLICIES";
})(EnablePolicyErrors || (exports.EnablePolicyErrors = EnablePolicyErrors = {}));
var DisablePolicyErrors;
(function (DisablePolicyErrors) {
    DisablePolicyErrors["DISABLE_POLICY_CANCELLED"] = "DISABLE_POLICY_CANCELLED";
    DisablePolicyErrors["NO_ENABLED_POLICIES"] = "NO_ENABLED_POLICIES";
})(DisablePolicyErrors || (exports.DisablePolicyErrors = DisablePolicyErrors = {}));
var GetToolPolicyErrors;
(function (GetToolPolicyErrors) {
    GetToolPolicyErrors["GET_TOOL_POLICY_CANCELLED"] = "GET_TOOL_POLICY_CANCELLED";
    GetToolPolicyErrors["NO_TOOLS_WITH_POLICIES"] = "NO_TOOLS_WITH_POLICIES";
})(GetToolPolicyErrors || (exports.GetToolPolicyErrors = GetToolPolicyErrors = {}));
var GetDelegateesErrors;
(function (GetDelegateesErrors) {
    GetDelegateesErrors["NO_DELEGATEES_FOUND"] = "NO_DELEGATEES_FOUND";
})(GetDelegateesErrors || (exports.GetDelegateesErrors = GetDelegateesErrors = {}));
var AddDelegateeErrors;
(function (AddDelegateeErrors) {
    AddDelegateeErrors["ADD_DELEGATEE_CANCELLED"] = "ADD_DELEGATEE_CANCELLED";
})(AddDelegateeErrors || (exports.AddDelegateeErrors = AddDelegateeErrors = {}));
var RemoveDelegateeErrors;
(function (RemoveDelegateeErrors) {
    RemoveDelegateeErrors["REMOVE_DELEGATEE_CANCELLED"] = "REMOVE_DELEGATEE_CANCELLED";
    RemoveDelegateeErrors["NO_DELEGATEES_FOUND"] = "NO_DELEGATEES_FOUND";
})(RemoveDelegateeErrors || (exports.RemoveDelegateeErrors = RemoveDelegateeErrors = {}));
var IsDelegateeErrors;
(function (IsDelegateeErrors) {
    IsDelegateeErrors["IS_DELEGATEE_CANCELLED"] = "IS_DELEGATEE_CANCELLED";
})(IsDelegateeErrors || (exports.IsDelegateeErrors = IsDelegateeErrors = {}));
var ManageDelegateesMenuErrors;
(function (ManageDelegateesMenuErrors) {
    ManageDelegateesMenuErrors["NO_MANAGE_DELEGATEES_ACTION_SELECTED"] = "NO_MANAGE_DELEGATEES_ACTION_SELECTED";
    ManageDelegateesMenuErrors["UNKNOWN_MANAGE_DELEGATEES_ACTION"] = "UNKNOWN_MANAGE_DELEGATEES_ACTION";
})(ManageDelegateesMenuErrors || (exports.ManageDelegateesMenuErrors = ManageDelegateesMenuErrors = {}));
var PermitToolForDelegateeErrors;
(function (PermitToolForDelegateeErrors) {
    PermitToolForDelegateeErrors["NO_TOOLS_FOUND"] = "NO_TOOLS_FOUND";
    PermitToolForDelegateeErrors["NO_DELEGATEES_FOUND"] = "NO_DELEGATEES_FOUND";
    PermitToolForDelegateeErrors["PERMIT_TOOL_FOR_DELEGATEE_CANCELLED"] = "PERMIT_TOOL_FOR_DELEGATEE_CANCELLED";
})(PermitToolForDelegateeErrors || (exports.PermitToolForDelegateeErrors = PermitToolForDelegateeErrors = {}));
var UnpermitToolForDelegateeErrors;
(function (UnpermitToolForDelegateeErrors) {
    UnpermitToolForDelegateeErrors["NO_PERMITTED_TOOLS"] = "NO_PERMITTED_TOOLS";
    UnpermitToolForDelegateeErrors["UNPERMIT_TOOL_FOR_DELEGATEE_CANCELLED"] = "UNPERMIT_TOOL_FOR_DELEGATEE_CANCELLED";
})(UnpermitToolForDelegateeErrors || (exports.UnpermitToolForDelegateeErrors = UnpermitToolForDelegateeErrors = {}));
var GetToolPolicyParameterErrors;
(function (GetToolPolicyParameterErrors) {
    GetToolPolicyParameterErrors["NO_TOOLS_WITH_POLICIES"] = "NO_TOOLS_WITH_POLICIES";
    GetToolPolicyParameterErrors["NO_DELEGATEES"] = "NO_DELEGATEES";
    GetToolPolicyParameterErrors["NO_PARAMETERS"] = "NO_PARAMETERS";
    GetToolPolicyParameterErrors["GET_CANCELLED"] = "GET_CANCELLED";
})(GetToolPolicyParameterErrors || (exports.GetToolPolicyParameterErrors = GetToolPolicyParameterErrors = {}));
var SetToolPolicyParameterErrors;
(function (SetToolPolicyParameterErrors) {
    SetToolPolicyParameterErrors["NO_TOOLS_WITH_POLICIES"] = "NO_TOOLS_WITH_POLICIES";
    SetToolPolicyParameterErrors["NO_DELEGATEES"] = "NO_DELEGATEES";
    SetToolPolicyParameterErrors["SET_CANCELLED"] = "SET_CANCELLED";
    SetToolPolicyParameterErrors["PARAMETER_EXISTS"] = "PARAMETER_EXISTS";
    SetToolPolicyParameterErrors["FAILED"] = "FAILED";
})(SetToolPolicyParameterErrors || (exports.SetToolPolicyParameterErrors = SetToolPolicyParameterErrors = {}));
var RemoveToolPolicyParameterErrors;
(function (RemoveToolPolicyParameterErrors) {
    RemoveToolPolicyParameterErrors["NO_TOOLS_WITH_POLICIES"] = "NO_TOOLS_WITH_POLICIES";
    RemoveToolPolicyParameterErrors["NO_DELEGATEES"] = "NO_DELEGATEES";
    RemoveToolPolicyParameterErrors["NO_PARAMETERS"] = "NO_PARAMETERS";
    RemoveToolPolicyParameterErrors["REMOVE_CANCELLED"] = "REMOVE_CANCELLED";
    RemoveToolPolicyParameterErrors["FAILED"] = "FAILED";
})(RemoveToolPolicyParameterErrors || (exports.RemoveToolPolicyParameterErrors = RemoveToolPolicyParameterErrors = {}));
var DelegateeErrors;
(function (DelegateeErrors) {
    DelegateeErrors["DELEGATEE_SELECTION_CANCELLED"] = "DELEGATEE_SELECTION_CANCELLED";
    DelegateeErrors["DELEGATEE_MISSING_PRIVATE_KEY"] = "DELEGATEE_MISSING_PRIVATE_KEY";
    DelegateeErrors["FAILED_TO_INITIALIZE_DELEGATEE"] = "FAILED_TO_INITIALIZE_DELEGATEE";
    DelegateeErrors["NO_DELEGATED_PKPS"] = "NO_DELEGATED_PKPS";
    DelegateeErrors["NO_TOOLS_WITH_POLICIES"] = "NO_TOOLS_WITH_POLICIES";
    DelegateeErrors["TOOL_SELECTION_CANCELLED"] = "TOOL_SELECTION_CANCELLED";
    DelegateeErrors["GET_TOOL_VIA_INTENT_CANCELLED"] = "GET_TOOL_VIA_INTENT_CANCELLED";
    DelegateeErrors["GET_INTENT_MATCHER_CANCELLED"] = "GET_INTENT_MATCHER_CANCELLED";
    DelegateeErrors["NO_TOOLS_AVAILABLE"] = "NO_TOOLS_AVAILABLE";
    DelegateeErrors["TOOL_PARAMS_CANCELLED"] = "TOOL_PARAMS_CANCELLED";
    DelegateeErrors["TOOL_PARAMS_INVALID"] = "TOOL_PARAMS_INVALID";
    DelegateeErrors["NO_CHAINS_CONFIGURED"] = "NO_CHAINS_CONFIGURED";
    DelegateeErrors["CHAIN_SELECTION_CANCELLED"] = "CHAIN_SELECTION_CANCELLED";
})(DelegateeErrors || (exports.DelegateeErrors = DelegateeErrors = {}));
