// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/core/law-cli-error/law-cli-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LawCliError = void 0;
class LawCliError extends Error {
    type;
    details;
    serializedDetails;
    constructor(type, message, details) {
        super(message);
        this.type = type;
        this.details = details;
        this.name = 'AwCliError';
        // Store a serialized version of details for better error logging
        this.serializedDetails = details
            ? JSON.stringify(details, (key, value) => {
                if (value instanceof Error) {
                    // Handle nested errors
                    return {
                        name: value.name,
                        message: value.message,
                        stack: value.stack,
                        ...(value instanceof LawCliError
                            ? {
                                type: value.type,
                                details: value.serializedDetails
                                    ? JSON.parse(value.serializedDetails)
                                    : undefined,
                            }
                            : {}),
                    };
                }
                return value;
            }, 2)
            : '';
    }
    // Override toJSON to provide better serialization
    toJSON() {
        return {
            name: this.name,
            type: this.type,
            message: this.message,
            details: this.serializedDetails
                ? JSON.parse(this.serializedDetails)
                : undefined,
            stack: this.stack,
        };
    }
}
exports.LawCliError = LawCliError;
