// @virtuals-protocol/game@5.7.2/dist/typechain-types/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirtualsToken__factory = exports.VirtualsPolicy__factory = exports.LitAgentWallet__factory = exports.ShortStrings__factory = exports.EIP712__factory = exports.IERC20__factory = exports.IERC20Permit__factory = exports.IERC20Metadata__factory = exports.ERC20Votes__factory = exports.ERC20Permit__factory = exports.ERC20__factory = exports.Pausable__factory = exports.IERC6372__factory = exports.IERC5805__factory = exports.IERC5267__factory = exports.IVotes__factory = exports.Ownable__factory = exports.factories = void 0;
exports.factories = __importStar(require("./factories"));
var Ownable__factory_1 = require("./factories/@openzeppelin/contracts/access/Ownable__factory");
Object.defineProperty(exports, "Ownable__factory", { enumerable: true, get: function () { return Ownable__factory_1.Ownable__factory; } });
var IVotes__factory_1 = require("./factories/@openzeppelin/contracts/governance/utils/IVotes__factory");
Object.defineProperty(exports, "IVotes__factory", { enumerable: true, get: function () { return IVotes__factory_1.IVotes__factory; } });
var IERC5267__factory_1 = require("./factories/@openzeppelin/contracts/interfaces/IERC5267__factory");
Object.defineProperty(exports, "IERC5267__factory", { enumerable: true, get: function () { return IERC5267__factory_1.IERC5267__factory; } });
var IERC5805__factory_1 = require("./factories/@openzeppelin/contracts/interfaces/IERC5805__factory");
Object.defineProperty(exports, "IERC5805__factory", { enumerable: true, get: function () { return IERC5805__factory_1.IERC5805__factory; } });
var IERC6372__factory_1 = require("./factories/@openzeppelin/contracts/interfaces/IERC6372__factory");
Object.defineProperty(exports, "IERC6372__factory", { enumerable: true, get: function () { return IERC6372__factory_1.IERC6372__factory; } });
var Pausable__factory_1 = require("./factories/@openzeppelin/contracts/security/Pausable__factory");
Object.defineProperty(exports, "Pausable__factory", { enumerable: true, get: function () { return Pausable__factory_1.Pausable__factory; } });
var ERC20__factory_1 = require("./factories/@openzeppelin/contracts/token/ERC20/ERC20__factory");
Object.defineProperty(exports, "ERC20__factory", { enumerable: true, get: function () { return ERC20__factory_1.ERC20__factory; } });
var ERC20Permit__factory_1 = require("./factories/@openzeppelin/contracts/token/ERC20/extensions/ERC20Permit__factory");
Object.defineProperty(exports, "ERC20Permit__factory", { enumerable: true, get: function () { return ERC20Permit__factory_1.ERC20Permit__factory; } });
var ERC20Votes__factory_1 = require("./factories/@openzeppelin/contracts/token/ERC20/extensions/ERC20Votes__factory");
Object.defineProperty(exports, "ERC20Votes__factory", { enumerable: true, get: function () { return ERC20Votes__factory_1.ERC20Votes__factory; } });
var IERC20Metadata__factory_1 = require("./factories/@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata__factory");
Object.defineProperty(exports, "IERC20Metadata__factory", { enumerable: true, get: function () { return IERC20Metadata__factory_1.IERC20Metadata__factory; } });
var IERC20Permit__factory_1 = require("./factories/@openzeppelin/contracts/token/ERC20/extensions/IERC20Permit__factory");
Object.defineProperty(exports, "IERC20Permit__factory", { enumerable: true, get: function () { return IERC20Permit__factory_1.IERC20Permit__factory; } });
var IERC20__factory_1 = require("./factories/@openzeppelin/contracts/token/ERC20/IERC20__factory");
Object.defineProperty(exports, "IERC20__factory", { enumerable: true, get: function () { return IERC20__factory_1.IERC20__factory; } });
var EIP712__factory_1 = require("./factories/@openzeppelin/contracts/utils/cryptography/EIP712__factory");
Object.defineProperty(exports, "EIP712__factory", { enumerable: true, get: function () { return EIP712__factory_1.EIP712__factory; } });
var ShortStrings__factory_1 = require("./factories/@openzeppelin/contracts/utils/ShortStrings__factory");
Object.defineProperty(exports, "ShortStrings__factory", { enumerable: true, get: function () { return ShortStrings__factory_1.ShortStrings__factory; } });
var LitAgentWallet__factory_1 = require("./factories/src/contracts/LitAgentWallet__factory");
Object.defineProperty(exports, "LitAgentWallet__factory", { enumerable: true, get: function () { return LitAgentWallet__factory_1.LitAgentWallet__factory; } });
var VirtualsPolicy__factory_1 = require("./factories/src/contracts/VirtualsPolicy__factory");
Object.defineProperty(exports, "VirtualsPolicy__factory", { enumerable: true, get: function () { return VirtualsPolicy__factory_1.VirtualsPolicy__factory; } });
var VirtualsToken__factory_1 = require("./factories/src/contracts/VirtualsToken__factory");
Object.defineProperty(exports, "VirtualsToken__factory", { enumerable: true, get: function () { return VirtualsToken__factory_1.VirtualsToken__factory; } });
