// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/main-menu/admin-menu/admin-settings/configure-signer/menu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleConfigureAdminSignerMenu = exports.AdminSignerType = exports.AdminConfigureSignerMenuChoice = void 0;
const tslib_1 = require("tslib");
const prompts_1 = tslib_1.__importDefault(require("prompts"));
const core_1 = require("../../../../core");
var AdminConfigureSignerMenuChoice;
(function (AdminConfigureSignerMenuChoice) {
    AdminConfigureSignerMenuChoice["UseEoa"] = "useEoa";
    AdminConfigureSignerMenuChoice["UseMultiSig"] = "useMultiSig";
    AdminConfigureSignerMenuChoice["UsePkp"] = "usePkp";
    AdminConfigureSignerMenuChoice["Back"] = "back";
})(AdminConfigureSignerMenuChoice || (exports.AdminConfigureSignerMenuChoice = AdminConfigureSignerMenuChoice = {}));
var AdminSignerType;
(function (AdminSignerType) {
    AdminSignerType["Eoa"] = "eoa";
    AdminSignerType["MultiSig"] = "multiSig";
    AdminSignerType["Pkp"] = "pkp";
})(AdminSignerType || (exports.AdminSignerType = AdminSignerType = {}));
const promptConfigureSignerMenu = async () => {
    const { action } = await (0, prompts_1.default)({
        type: 'select',
        name: 'action',
        message: 'What would you like to do?',
        choices: [
            {
                title: 'Use EOA',
                value: AdminConfigureSignerMenuChoice.UseEoa,
            },
            {
                title: 'Use MultiSig',
                value: AdminConfigureSignerMenuChoice.UseMultiSig,
                disabled: true,
            },
            {
                title: 'Use Pkp',
                value: AdminConfigureSignerMenuChoice.UsePkp,
                disabled: true,
            },
            { title: 'Back', value: AdminConfigureSignerMenuChoice.Back },
        ],
    });
    if (!action) {
        throw new core_1.LawCliError(core_1.GeneralErrors.NO_ACTION_SELECTED, 'No admin menu action selected. Please select an action to continue.');
    }
    return action;
};
const handleConfigureAdminSignerMenu = async () => {
    try {
        return promptConfigureSignerMenu();
    }
    catch (error) {
        if (error instanceof core_1.LawCliError) {
            if (error.type === core_1.GeneralErrors.NO_ACTION_SELECTED) {
                core_1.logger.error(error.message);
                return await (0, exports.handleConfigureAdminSignerMenu)();
            }
        }
        throw error;
    }
};
exports.handleConfigureAdminSignerMenu = handleConfigureAdminSignerMenu;
