// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/main-menu/delegatee-menu/delegatee-settings/configure-signer/use-eoa.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleUseEoaForDelegatee = void 0;
const tslib_1 = require("tslib");
const prompts_1 = tslib_1.__importDefault(require("prompts"));
const ethers_1 = require("ethers");
const core_1 = require("../../../../core");
const menu_1 = require("./menu");
const delegatee_1 = require("../../delegatee");
const promptSelectDelegatee = async (localStorage) => {
    const delegatees = getStoredDelegatees(localStorage);
    const addresses = Object.keys(delegatees);
    if (addresses.length === 0) {
        return null;
    }
    const choices = [
        {
            title: 'Use new private key',
            value: 'new',
        },
        ...addresses.map((addr) => ({
            title: addr,
            value: addr,
        })),
    ];
    const { selection } = await (0, prompts_1.default)({
        type: 'select',
        name: 'selection',
        message: 'Select a delegatee wallet or add a new one:',
        choices,
    });
    if (!selection) {
        throw new core_1.LawCliError(core_1.DelegateeErrors.DELEGATEE_SELECTION_CANCELLED, 'Delegatee selection cancelled.');
    }
    if (selection === 'new') {
        return null;
    }
    return {
        address: selection,
        privateKey: delegatees[selection].privateKey,
    };
};
const promptPrivateKey = async () => {
    const { privateKey } = await (0, prompts_1.default)({
        type: 'password',
        name: 'privateKey',
        message: 'Enter your private key:',
        validate: (value) => {
            // Basic validation for private key format (0x followed by 64 hex characters)
            const isValidFormat = /^0x[0-9a-fA-F]{64}$/.test(value);
            return isValidFormat
                ? true
                : 'Please enter a valid private key (0x followed by 64 hex characters)';
        },
    });
    if (!privateKey) {
        throw new core_1.LawCliError(core_1.DelegateeErrors.DELEGATEE_MISSING_PRIVATE_KEY, 'No private key provided. Operation cancelled.');
    }
    return privateKey;
};
const getStoredDelegatees = (localStorage) => {
    const storedData = localStorage.getItem(core_1.StorageKeys.DELEGATEE_STORAGE);
    return storedData ? JSON.parse(storedData) : {};
};
const saveDelegatee = (localStorage, address, privateKey) => {
    const delegatees = getStoredDelegatees(localStorage);
    delegatees[address] = {
        privateKey,
    };
    localStorage.setItem(core_1.StorageKeys.DELEGATEE_STORAGE, JSON.stringify(delegatees));
};
const handleUseEoaForDelegatee = async (localStorage) => {
    try {
        // First check if we have any stored delegatees
        const existingDelegatee = await promptSelectDelegatee(localStorage);
        let privateKey;
        let address;
        if (existingDelegatee) {
            privateKey = existingDelegatee.privateKey;
            address = existingDelegatee.address;
        }
        else {
            privateKey = await promptPrivateKey();
            // Create wallet to get address
            const wallet = new ethers_1.ethers.Wallet(privateKey);
            address = wallet.address;
            // Save new delegatee
            saveDelegatee(localStorage, address, privateKey);
        }
        // Set the current active delegatee
        localStorage.setItem(core_1.StorageKeys.DELEGATEE_ACTIVE_ADDRESS, address);
        localStorage.setItem(core_1.StorageKeys.DELEGATEE_SIGNER_TYPE, menu_1.DelegateeSignerType.Eoa);
        const litNetwork = await (0, core_1.getLitNetwork)(localStorage);
        const awDelegatee = await delegatee_1.Delegatee.create(litNetwork, privateKey);
        core_1.logger.success(`EOA signer configured successfully with address: ${address}`);
        return awDelegatee;
    }
    catch (error) {
        if (error instanceof core_1.LawCliError) {
            if (error.type === core_1.DelegateeErrors.DELEGATEE_MISSING_PRIVATE_KEY ||
                error.type === core_1.DelegateeErrors.FAILED_TO_INITIALIZE_DELEGATEE ||
                error.type === core_1.DelegateeErrors.DELEGATEE_SELECTION_CANCELLED) {
                core_1.logger.error(error.message);
                return await (0, exports.handleUseEoaForDelegatee)(localStorage);
            }
        }
        throw error;
    }
};
exports.handleUseEoaForDelegatee = handleUseEoaForDelegatee;
