// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/main-menu/main-menu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleMainMenu = exports.MainMenuChoice = void 0;
const tslib_1 = require("tslib");
const prompts_1 = tslib_1.__importDefault(require("prompts"));
const core_1 = require("../core");
var MainMenuChoice;
(function (MainMenuChoice) {
    MainMenuChoice["AdminMenu"] = "adminMenu";
    MainMenuChoice["DelegateeMenu"] = "delegateeMenu";
    MainMenuChoice["CliSettings"] = "cliSettings";
})(MainMenuChoice || (exports.MainMenuChoice = MainMenuChoice = {}));
const promptMainMenu = async () => {
    const { action } = await (0, prompts_1.default)({
        type: 'select',
        name: 'action',
        message: 'What would you like to do?',
        choices: [
            {
                title: 'Admin Menu',
                value: MainMenuChoice.AdminMenu,
            },
            { title: 'Delegatee Menu', value: MainMenuChoice.DelegateeMenu },
            { title: 'Change CLI Settings', value: MainMenuChoice.CliSettings },
        ],
    });
    if (!action) {
        throw new core_1.LawCliError(core_1.MainMenuErrors.NO_MAIN_MENU_ACTION_SELECTED, 'No main menu action selected. Please select an action to continue.');
    }
    return action;
};
const handleMainMenu = async () => {
    try {
        return promptMainMenu();
    }
    catch (error) {
        if (error instanceof core_1.LawCliError) {
            if (error.type === core_1.MainMenuErrors.NO_MAIN_MENU_ACTION_SELECTED) {
                core_1.logger.error(error.message);
                return await (0, exports.handleMainMenu)();
            }
        }
        throw error;
    }
};
exports.handleMainMenu = handleMainMenu;
