// @lit-protocol/wrapped-keys/src/lib/lit-actions-client/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LIT_ACTION_CID_REPOSITORY_COMMON = exports.LIT_ACTION_CID_REPOSITORY = void 0;
const LIT_ACTION_CID_REPOSITORY = Object.freeze({
    signTransaction: Object.freeze({
        evm: 'QmRpAgGKEmgeBRhqdC2EH17QUt6puwsbm8Z2nNneVN4uJG',
        solana: 'QmR1nPG2tnmC72zuCEMZUZrrMEkbDiMPNHW45Dsm2n7xnk',
    }),
    signMessage: Object.freeze({
        evm: 'QmXi9iqJvXrHoUGSo5WREonrruDhzQ7cFr7Cry3wX2hmue',
        solana: 'QmcEJGVqRYtVukjm2prCPT7Fs66GpaqZwmZoxEHMHor6Jz',
    }),
    generateEncryptedKey: Object.freeze({
        evm: 'QmeD6NYCWhUCLgxgpkgSguaKjwnpCnJ6Yf8SdsyPpK4eKK',
        solana: 'QmPkVD3hEjMi1T5zQuvdrFCXaGTEMHNdAhAL4WHkqxijrQ',
    }),
    exportPrivateKey: Object.freeze({
        evm: 'QmUJ74pTUqeeHzDGdfwCph1vJVNJ1rRzJdvMiTjS1BMwYj',
        solana: 'QmUJ74pTUqeeHzDGdfwCph1vJVNJ1rRzJdvMiTjS1BMwYj',
    }),
});
exports.LIT_ACTION_CID_REPOSITORY = LIT_ACTION_CID_REPOSITORY;
const LIT_ACTION_CID_REPOSITORY_COMMON = Object.freeze({
    batchGenerateEncryptedKeys: 'QmR8Zs7ctSEctxBrSnAYhMXFXCC1ub8K1xvMn5Js3NCSAA',
});
exports.LIT_ACTION_CID_REPOSITORY_COMMON = LIT_ACTION_CID_REPOSITORY_COMMON;
