// @lit-protocol/wrapped-keys/src/lib/lit-actions-client/export-private-key.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.exportPrivateKeyWithLitAction = exportPrivateKeyWithLitAction;
const constants_1 = require("@lit-protocol/constants");
const utils_1 = require("./utils");
async function exportPrivateKeyWithLitAction(args) {
    const { accessControlConditions, litNodeClient, pkpSessionSigs, litActionCode, litActionIpfsCid, storedKeyMetadata, } = args;
    const { pkpAddress, ciphertext, dataToEncryptHash, ...storeKeyMetadataMinusEncryptedAndPkp } = storedKeyMetadata;
    const result = await litNodeClient.executeJs({
        sessionSigs: pkpSessionSigs,
        code: litActionCode,
        ipfsId: litActionIpfsCid,
        jsParams: {
            pkpAddress,
            ciphertext,
            dataToEncryptHash,
            accessControlConditions,
        },
        ipfsOptions: {
            overwriteCode: constants_1.GLOBAL_OVERWRITE_IPFS_CODE_BY_NETWORK[litNodeClient.config.litNetwork],
        },
    });
    const decryptedPrivateKey = (0, utils_1.postLitActionValidation)(result);
    return {
        decryptedPrivateKey,
        pkpAddress,
        ...storeKeyMetadataMinusEncryptedAndPkp,
    };
}
