// @lit-protocol/wrapped-keys/src/lib/lit-actions-client/sign-message.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signMessageWithLitAction = signMessageWithLitAction;
const constants_1 = require("@lit-protocol/constants");
const utils_1 = require("./utils");
async function signMessageWithLitAction(args) {
    const { accessControlConditions, litNodeClient, messageToSign, pkpSessionSigs, litActionIpfsCid, litActionCode, storedKeyMetadata, } = args;
    const { pkpAddress, ciphertext, dataToEncryptHash } = storedKeyMetadata;
    const result = await litNodeClient.executeJs({
        sessionSigs: pkpSessionSigs,
        ipfsId: litActionIpfsCid,
        code: litActionCode,
        jsParams: {
            pkpAddress,
            ciphertext,
            dataToEncryptHash,
            messageToSign,
            accessControlConditions,
        },
        ipfsOptions: {
            overwriteCode: constants_1.GLOBAL_OVERWRITE_IPFS_CODE_BY_NETWORK[litNodeClient.config.litNetwork],
        },
    });
    return (0, utils_1.postLitActionValidation)(result);
}
