// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/main-menu/cli-settings/change-lit-network.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleChangeLitNetwork = void 0;
const tslib_1 = require("tslib");
const prompts_1 = tslib_1.__importDefault(require("prompts"));
const core_1 = require("../../core");
const promptChangeLitNetwork = async () => {
    const { network } = await (0, prompts_1.default)({
        type: 'select',
        name: 'network',
        message: 'Select a Lit network:',
        choices: [
            {
                title: 'Datil Dev',
                description: 'Development network',
                value: 'datil-dev',
            },
            {
                title: 'Datil Test',
                description: 'Pre-production test network',
                value: 'datil-test',
            },
            {
                title: 'Datil',
                description: 'Production network',
                value: 'datil',
            },
        ],
    });
    if (!network) {
        throw new core_1.LawCliError(core_1.ChangeLitNetworkErrors.NO_LIT_NETWORK_SELECTED, 'No Lit network selected. Please select a network to continue.');
    }
    return network;
};
const handleChangeLitNetwork = async (localStorage) => {
    try {
        const selectedLitNetwork = await promptChangeLitNetwork();
        localStorage.setItem(core_1.StorageKeys.LIT_NETWORK, selectedLitNetwork);
        return selectedLitNetwork;
    }
    catch (error) {
        if (error instanceof core_1.LawCliError) {
            if (error.type === core_1.ChangeLitNetworkErrors.NO_LIT_NETWORK_SELECTED) {
                core_1.logger.error(error.message);
                return await (0, exports.handleChangeLitNetwork)(localStorage);
            }
        }
        throw error;
    }
};
exports.handleChangeLitNetwork = handleChangeLitNetwork;
