// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/main-menu/cli-settings/menu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleCliSettingsMenu = exports.CliSettingsMenuChoice = void 0;
const tslib_1 = require("tslib");
const prompts_1 = tslib_1.__importDefault(require("prompts"));
const core_1 = require("../../core");
var CliSettingsMenuChoice;
(function (CliSettingsMenuChoice) {
    CliSettingsMenuChoice["ChangeLitNetwork"] = "changeLitNetwork";
    CliSettingsMenuChoice["ManageRpcs"] = "manageRpcs";
    CliSettingsMenuChoice["Back"] = "back";
})(CliSettingsMenuChoice || (exports.CliSettingsMenuChoice = CliSettingsMenuChoice = {}));
const promptCliSettingsMenu = async () => {
    const { action } = await (0, prompts_1.default)({
        type: 'select',
        name: 'action',
        message: 'What would you like to do?',
        choices: [
            {
                title: 'Change Lit Network',
                value: CliSettingsMenuChoice.ChangeLitNetwork,
            },
            { title: 'Manage RPCs', value: CliSettingsMenuChoice.ManageRpcs },
            { title: 'Back', value: CliSettingsMenuChoice.Back },
        ],
    });
    if (!action) {
        throw new core_1.LawCliError(core_1.CliSettingsMenuErrors.NO_CLI_SETTINGS_ACTION_SELECTED, 'No CLI settings action selected. Please select an action to continue.');
    }
    return action;
};
const handleCliSettingsMenu = async () => {
    try {
        return promptCliSettingsMenu();
    }
    catch (error) {
        if (error instanceof core_1.LawCliError) {
            if (error.type === core_1.CliSettingsMenuErrors.NO_CLI_SETTINGS_ACTION_SELECTED) {
                core_1.logger.error(error.message);
                return await (0, exports.handleCliSettingsMenu)();
            }
        }
        throw error;
    }
};
exports.handleCliSettingsMenu = handleCliSettingsMenu;
