// @lit-protocol/wrapped-keys/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = exports.api = exports.constants = void 0;
const api_1 = require("./lib/api");
const constants_1 = require("./lib/constants");
const code_repository_1 = require("./lib/lit-actions-client/code-repository");
exports.constants = {
    CHAIN_ETHEREUM: constants_1.CHAIN_ETHEREUM,
    LIT_PREFIX: constants_1.LIT_PREFIX,
    NETWORK_EVM: constants_1.NETWORK_EVM,
    NETWORK_SOLANA: constants_1.NETWORK_SOLANA,
    KEYTYPE_K256: constants_1.KEYTYPE_K256,
    KEYTYPE_ED25519: constants_1.KEYTYPE_ED25519,
};
exports.api = {
    exportPrivateKey: api_1.exportPrivateKey,
    generatePrivateKey: api_1.generatePrivateKey,
    getEncryptedKey: api_1.getEncryptedKey,
    listEncryptedKeyMetadata: api_1.listEncryptedKeyMetadata,
    importPrivateKey: api_1.importPrivateKey,
    signMessageWithEncryptedKey: api_1.signMessageWithEncryptedKey,
    signTransactionWithEncryptedKey: api_1.signTransactionWithEncryptedKey,
    storeEncryptedKey: api_1.storeEncryptedKey,
    storeEncryptedKeyBatch: api_1.storeEncryptedKeyBatch,
    batchGeneratePrivateKeys: api_1.batchGeneratePrivateKeys,
};
exports.config = {
    setLitActionsCode: code_repository_1.setLitActionsCode,
    setLitActionsCodeCommon: code_repository_1.setLitActionsCodeCommon,
};
