// @lit-protocol/aw-tool-enso@0.1.0-19/dist/src/lib/lit-actions/utils/get-route.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRoute = void 0;
/**
 * Retrieves the best route from tokenIn to tokenOut through Enso
 * @param {EnsoClient} ensoClient - Enso API Client
 * @param {number} chainId - ChainID of blockchain network
 * @param {string} fromAddress - Address from whom the transaction will be executed
 * @param {string} tokenIn - The token to route away from
 * @param {string} amountIn - Amount of tokenIn
 * @param {string} tokenOut - The token to route to
 * @returns {Promise<{bestQuote: any;bestFee: number;amountOutMin: any;}>} The best quote and fee tier.
 */
const getRoute = async (ensoClient, chainId, fromAddress, tokenIn, amountIn, tokenOut) => {
    const routeParams = {
        chainId,
        tokenIn: tokenIn,
        tokenOut: tokenOut,
        amountIn: amountIn,
        fromAddress: fromAddress,
        receiver: fromAddress,
        spender: fromAddress,
    };
    console.log('Fetching the best route through Enso...');
    const routeData = await ensoClient.getRouterData(routeParams);
    return routeData;
    //return {
    //  ...routeData,
    //  tx: {
    //    to: routeData.tx.to,
    //    data: routeData.tx.data,
    //    value: routeData.tx.value,
    //    gasLimit: routeData.gas,
    //    maxFeePerGas: gasData.maxFeePerGas,
    //    maxPriorityFeePerGas: gasData.maxPriorityFeePerGas,
    //    nonce: gasData.nonce,
    //    chainId: chainId.toString(),
    //    type: 2,
    //  },
    //};
};
exports.getRoute = getRoute;
