// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/main-menu/admin-menu/manage-tools/disable-tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleDisableTool = void 0;
const tslib_1 = require("tslib");
const prompts_1 = tslib_1.__importDefault(require("prompts"));
const core_1 = require("../../../core");
/**
 * Prompts the user to select a tool to disable from the list of registered tools.
 * This function filters and presents only the tools that are currently enabled.
 *
 * @param registeredTools - Object containing information about registered tools
 * @returns The IPFS CID and name of the selected tool
 * @throws LawCliError - If no enabled tools are found or the user cancels the selection
 */
const promptSelectToolToDisable = async (registeredTools) => {
    // Combine tools with and without policies
    const allTools = {
        ...registeredTools.toolsWithPolicies,
        ...registeredTools.toolsWithoutPolicies,
    };
    // Filter for enabled tools
    const enabledTools = Object.entries(allTools).filter(([_, tool]) => tool.toolEnabled);
    if (enabledTools.length === 0) {
        throw new core_1.LawCliError(core_1.DisableToolErrors.NO_ENABLED_TOOLS, 'No enabled tools found to disable.');
    }
    const { tool } = await (0, prompts_1.default)({
        type: 'select',
        name: 'tool',
        message: 'Select a tool to disable:',
        choices: enabledTools.map(([ipfsCid, tool]) => ({
            title: tool.name,
            description: tool.description,
            value: { ipfsCid, name: tool.name },
        })),
    });
    if (!tool) {
        throw new core_1.LawCliError(core_1.DisableToolErrors.DISABLE_TOOL_CANCELLED, 'Tool disabling cancelled.');
    }
    return tool;
};
const handleDisableTool = async (admin, pkp) => {
    try {
        const registeredTools = await admin.awAdmin.getRegisteredToolsAndDelegateesForPkp(pkp.info.tokenId);
        const selectedTool = await promptSelectToolToDisable(registeredTools);
        await admin.awAdmin.disableTool(pkp.info.tokenId, selectedTool.ipfsCid);
        core_1.logger.success(`Tool ${selectedTool.name} disabled successfully.`);
    }
    catch (error) {
        if (error instanceof core_1.LawCliError) {
            if (error.type === core_1.DisableToolErrors.DISABLE_TOOL_CANCELLED ||
                error.type === core_1.DisableToolErrors.NO_ENABLED_TOOLS) {
                core_1.logger.error(error.message);
                return;
            }
        }
        throw error;
    }
};
exports.handleDisableTool = handleDisableTool;
