// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/main-menu/admin-menu/manage-tools/enable-tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleEnableTool = void 0;
const tslib_1 = require("tslib");
const prompts_1 = tslib_1.__importDefault(require("prompts"));
const core_1 = require("../../../core");
/**
 * Prompts the user to select a tool to enable from the list of registered tools.
 * This function filters and presents only the tools that are currently disabled.
 *
 * @param registeredTools - Object containing information about registered tools
 * @returns The IPFS CID and name of the selected tool
 * @throws LawCliError - If no disabled tools are found or the user cancels the selection
 */
const promptSelectToolToEnable = async (registeredTools) => {
    // Combine tools with and without policies
    const allTools = {
        ...registeredTools.toolsWithPolicies,
        ...registeredTools.toolsWithoutPolicies,
    };
    // Filter for disabled tools
    const disabledTools = Object.entries(allTools).filter(([_, tool]) => !tool.toolEnabled);
    if (disabledTools.length === 0) {
        throw new core_1.LawCliError(core_1.EnableToolErrors.NO_DISABLED_TOOLS, 'No disabled tools found to enable.');
    }
    const { tool } = await (0, prompts_1.default)({
        type: 'select',
        name: 'tool',
        message: 'Select a tool to enable:',
        choices: disabledTools.map(([ipfsCid, tool]) => ({
            title: tool.name,
            description: tool.description,
            value: { ipfsCid, name: tool.name },
        })),
    });
    if (!tool) {
        throw new core_1.LawCliError(core_1.EnableToolErrors.ENABLE_TOOL_CANCELLED, 'Tool enabling cancelled.');
    }
    return tool;
};
const handleEnableTool = async (admin, pkp) => {
    try {
        const registeredTools = await admin.awAdmin.getRegisteredToolsAndDelegateesForPkp(pkp.info.tokenId);
        const selectedTool = await promptSelectToolToEnable(registeredTools);
        await admin.awAdmin.enableTool(pkp.info.tokenId, selectedTool.ipfsCid);
        core_1.logger.success(`Tool ${selectedTool.name} enabled successfully.`);
    }
    catch (error) {
        if (error instanceof core_1.LawCliError) {
            if (error.type === core_1.EnableToolErrors.ENABLE_TOOL_CANCELLED ||
                error.type === core_1.EnableToolErrors.NO_DISABLED_TOOLS) {
                core_1.logger.error(error.message);
                return;
            }
        }
        throw error;
    }
};
exports.handleEnableTool = handleEnableTool;
