// @lit-protocol/aw-tool-sign-eddsa/dist/src/lib/policy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignEddsaPolicy = void 0;
const zod_1 = require("zod");
const ethers_1 = require("ethers");
/**
 * Schema for validating a SignEddsa policy.
 * Ensures the policy has the correct structure and valid values.
 */
const policySchema = zod_1.z.object({
    /** The type of policy, must be `SignEddsa`. */
    type: zod_1.z.literal('SignEddsa'),
    /** The version of the policy. */
    version: zod_1.z.string(),
    /** Array of allowed message prefixes. */
    allowedPrefixes: zod_1.z.array(zod_1.z.string())
});
/**
 * Encodes a SignEddsa policy into a format suitable for on-chain storage.
 * @param policy - The SignEddsa policy to encode.
 * @returns The encoded policy as a hex string.
 * @throws If the policy does not conform to the schema.
 */
function encodePolicy(policy) {
    // Validate the policy against the schema
    policySchema.parse(policy);
    return ethers_1.ethers.utils.defaultAbiCoder.encode(['tuple(string[] allowedPrefixes)'], [policy]);
}
/**
 * Decodes a SignEddsa policy from its on-chain encoded format.
 * @param encodedPolicy - The encoded policy as a hex string.
 * @returns The decoded SignEddsa policy.
 * @throws If the encoded policy is invalid or does not conform to the schema.
 */
function decodePolicy(encodedPolicy) {
    const decoded = ethers_1.ethers.utils.defaultAbiCoder.decode(['tuple(string[] allowedPrefixes)'], encodedPolicy)[0];
    const policy = {
        type: 'SignEddsa',
        version: '1.0.0',
        allowedPrefixes: decoded.allowedPrefixes
    };
    return policySchema.parse(policy);
}
/**
 * Utility object for working with SignEddsa policies.
 * Includes the schema, encoding, and decoding functions.
 */
exports.SignEddsaPolicy = {
    /** The type of the policy. */
    type: {},
    /** The version of the policy. */
    version: '1.0.0',
    /** The schema for validating SignEddsa policies. */
    schema: policySchema,
    /** Encodes a SignEddsa policy into a format suitable for on-chain storage. */
    encode: encodePolicy,
    /** Decodes a SignEddsa policy from its on-chain encoded format. */
    decode: decodePolicy,
};
