// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/core/storage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalStorage = exports.StorageKeys = void 0;
const node_localstorage_1 = require("node-localstorage");
var StorageKeys;
(function (StorageKeys) {
    StorageKeys["LIT_NETWORK"] = "litNetwork";
    StorageKeys["RPCS"] = "rpcs";
    StorageKeys["ADMIN_SIGNER_TYPE"] = "adminSignerType";
    StorageKeys["ADMIN_STORAGE"] = "adminStorage";
    StorageKeys["ADMIN_ACTIVE_ADDRESS"] = "adminActiveAddress";
    StorageKeys["DELEGATEE_STORAGE"] = "delegateeStorage";
    StorageKeys["DELEGATEE_ACTIVE_ADDRESS"] = "delegateeActiveAddress";
    StorageKeys["DELEGATEE_SIGNER_TYPE"] = "delegateeSignerType";
})(StorageKeys || (exports.StorageKeys = StorageKeys = {}));
class LocalStorage {
    storage;
    constructor(storageFilePath) {
        this.storage = new node_localstorage_1.LocalStorage(storageFilePath);
    }
    getItem(key) {
        return this.storage.getItem(key);
    }
    setItem(key, value) {
        this.storage.setItem(key, value);
    }
    removeItem(key) {
        this.storage.removeItem(key);
    }
    clear() {
        this.storage.clear();
    }
}
exports.LocalStorage = LocalStorage;
