// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/main-menu/admin-menu/manage-policies/disable-policy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleDisablePolicy = void 0;
const tslib_1 = require("tslib");
const prompts_1 = tslib_1.__importDefault(require("prompts"));
const core_1 = require("../../../core");
const promptSelectToolWithEnabledPolicy = async (registeredTools) => {
    // Get only tools with policies
    const toolsWithPolicies = registeredTools.toolsWithPolicies;
    // Filter for tools with enabled policies
    const toolsWithEnabledPolicies = Object.entries(toolsWithPolicies).filter(([_, tool]) => Object.values(tool.delegateePolicies).some((policy) => policy.policyEnabled));
    if (toolsWithEnabledPolicies.length === 0) {
        throw new core_1.LawCliError(core_1.DisablePolicyErrors.NO_ENABLED_POLICIES, 'No enabled policies found.');
    }
    const { tool } = await (0, prompts_1.default)({
        type: 'select',
        name: 'tool',
        message: 'Select a tool to disable policy for:',
        choices: toolsWithEnabledPolicies.map(([ipfsCid, tool]) => ({
            title: tool.name,
            description: `${Object.values(tool.delegateePolicies).filter((policy) => policy.policyEnabled).length} enabled policies`,
            value: {
                ipfsCid,
                name: tool.name,
                delegateePolicies: tool.delegateePolicies,
            },
        })),
    });
    if (!tool) {
        throw new core_1.LawCliError(core_1.DisablePolicyErrors.DISABLE_POLICY_CANCELLED, 'Tool selection cancelled.');
    }
    return tool;
};
const promptSelectDelegatee = async (delegateePolicies) => {
    // Filter for delegatees with enabled policies
    const enabledPolicies = Object.entries(delegateePolicies).filter(([_, policy]) => policy.policyEnabled);
    const { delegatee } = await (0, prompts_1.default)({
        type: 'select',
        name: 'delegatee',
        message: 'Select a delegatee to disable policy for:',
        choices: enabledPolicies.map(([address, policy]) => ({
            title: address,
            description: `Policy: ${policy.policyIpfsCid}`,
            value: address,
        })),
    });
    if (!delegatee) {
        throw new core_1.LawCliError(core_1.DisablePolicyErrors.DISABLE_POLICY_CANCELLED, 'Delegatee selection cancelled.');
    }
    return delegatee;
};
const promptConfirmDisable = async (toolName, delegatee) => {
    const { confirmed } = await (0, prompts_1.default)({
        type: 'confirm',
        name: 'confirmed',
        message: `Are you sure you want to disable the policy for tool "${toolName}" and delegatee "${delegatee}"?`,
        initial: false,
    });
    if (!confirmed) {
        throw new core_1.LawCliError(core_1.DisablePolicyErrors.DISABLE_POLICY_CANCELLED, 'Policy disable cancelled.');
    }
    return confirmed;
};
const handleDisablePolicy = async (admin, pkp) => {
    try {
        const registeredTools = await admin.awAdmin.getRegisteredToolsAndDelegateesForPkp(pkp.info.tokenId);
        const selectedTool = await promptSelectToolWithEnabledPolicy(registeredTools);
        const selectedDelegatee = await promptSelectDelegatee(selectedTool.delegateePolicies);
        await promptConfirmDisable(selectedTool.name, selectedDelegatee);
        await admin.awAdmin.disableToolPolicyForDelegatee(pkp.info.tokenId, selectedTool.ipfsCid, selectedDelegatee);
        core_1.logger.success(`Policy disabled successfully for tool ${selectedTool.name} and delegatee ${selectedDelegatee}.`);
    }
    catch (error) {
        if (error instanceof core_1.LawCliError) {
            if (error.type === core_1.DisablePolicyErrors.DISABLE_POLICY_CANCELLED ||
                error.type === core_1.DisablePolicyErrors.NO_ENABLED_POLICIES) {
                core_1.logger.error(error.message);
                return;
            }
        }
        throw error;
    }
};
exports.handleDisablePolicy = handleDisablePolicy;
