// @lit-protocol/law-cli@0.1.0-19/dist/src/lib/main-menu/admin-menu/manage-policies/remove-tool-policy-parameter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleRemoveToolPolicyParameter = void 0;
const tslib_1 = require("tslib");
const prompts_1 = tslib_1.__importDefault(require("prompts"));
const core_1 = require("../../../core");
const promptSelectToolForPolicyParameter = async (registeredTools) => {
    const choices = Object.values(registeredTools.toolsWithPolicies).map((tool) => ({
        title: tool.name,
        description: `${Object.keys(tool.delegateePolicies).length} policies`,
        value: tool,
    }));
    const { tool } = await (0, prompts_1.default)({
        type: 'select',
        name: 'tool',
        message: 'Select a tool to remove policy parameters from:',
        choices,
    });
    if (!tool) {
        throw new core_1.LawCliError(core_1.RemoveToolPolicyParameterErrors.REMOVE_CANCELLED, 'Tool policy parameter removal cancelled.');
    }
    return tool;
};
const promptSelectDelegateeForPolicyParameter = async (delegatees) => {
    if (delegatees.length === 0) {
        throw new core_1.LawCliError(core_1.RemoveToolPolicyParameterErrors.NO_DELEGATEES, 'No delegatees found.');
    }
    const { delegatee } = await (0, prompts_1.default)({
        type: 'select',
        name: 'delegatee',
        message: 'Select a delegatee to remove policy parameters from:',
        choices: delegatees.map((delegatee) => ({
            title: delegatee,
            value: delegatee,
        })),
    });
    if (!delegatee) {
        throw new core_1.LawCliError(core_1.RemoveToolPolicyParameterErrors.REMOVE_CANCELLED, 'Tool policy parameter removal cancelled.');
    }
    return delegatee;
};
const promptParametersToRemove = async (admin, pkp, tool, delegatee) => {
    const parameters = await admin.awAdmin.getAllToolPolicyParametersForDelegatee(pkp.info.tokenId, tool.ipfsCid, delegatee);
    if (!parameters || parameters.length === 0) {
        throw new core_1.LawCliError(core_1.RemoveToolPolicyParameterErrors.NO_PARAMETERS, 'No parameters found to remove.');
    }
    const choices = parameters.map((param) => ({
        title: param.name,
        value: param.name,
    }));
    const { selectedParameters } = await (0, prompts_1.default)({
        type: 'multiselect',
        name: 'selectedParameters',
        message: 'Select parameters to remove (Space to select):',
        choices,
        min: 1,
        instructions: false,
    });
    if (!selectedParameters || selectedParameters.length === 0) {
        throw new core_1.LawCliError(core_1.RemoveToolPolicyParameterErrors.REMOVE_CANCELLED, 'Tool policy parameter removal cancelled.');
    }
    return selectedParameters;
};
const removeToolPolicyParameters = async (admin, pkp, tool, delegatee, parameterNames) => {
    try {
        await admin.awAdmin.removeToolPolicyParametersForDelegatee(pkp.info.tokenId, tool.ipfsCid, delegatee, parameterNames);
        core_1.logger.success(`Successfully removed ${parameterNames.length} policy parameter(s) for tool ${tool.name} and delegatee ${delegatee}`);
    }
    catch (err) {
        throw new core_1.LawCliError(core_1.RemoveToolPolicyParameterErrors.FAILED, `Failed to remove tool policy parameters: ${err.message}`);
    }
};
const handleRemoveToolPolicyParameter = async (admin, pkp) => {
    try {
        const registeredTools = await admin.awAdmin.getRegisteredToolsAndDelegateesForPkp(pkp.info.tokenId);
        if (registeredTools === null ||
            Object.keys(registeredTools.toolsWithPolicies).length === 0) {
            throw new core_1.LawCliError(core_1.RemoveToolPolicyParameterErrors.NO_TOOLS_WITH_POLICIES, 'No tools with policies found.');
        }
        const selectedTool = await promptSelectToolForPolicyParameter(registeredTools);
        const selectedDelegatee = await promptSelectDelegateeForPolicyParameter(selectedTool.delegatees);
        const parameterNames = await promptParametersToRemove(admin, pkp, selectedTool, selectedDelegatee);
        await removeToolPolicyParameters(admin, pkp, selectedTool, selectedDelegatee, parameterNames);
    }
    catch (error) {
        if (error instanceof core_1.LawCliError) {
            if (error.type === core_1.RemoveToolPolicyParameterErrors.NO_TOOLS_WITH_POLICIES ||
                error.type === core_1.RemoveToolPolicyParameterErrors.NO_DELEGATEES ||
                error.type === core_1.RemoveToolPolicyParameterErrors.NO_PARAMETERS ||
                error.type === core_1.RemoveToolPolicyParameterErrors.REMOVE_CANCELLED ||
                error.type === core_1.RemoveToolPolicyParameterErrors.FAILED) {
                core_1.logger.error(error.message);
                return;
            }
        }
        throw error;
    }
};
exports.handleRemoveToolPolicyParameter = handleRemoveToolPolicyParameter;
