// @lit-protocol/aw-tool-jupiter-swap/dist/src/lib/lit-actions/utils/swap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getJupiterSwapTransaction = getJupiterSwapTransaction;
/// <reference lib="dom" />
const web3_js_1 = require("@solana/web3.js");
async function getJupiterSwapTransaction({ quoteResponse, userPublicKey }) {
    const response = await fetch('https://quote-api.jup.ag/v6/swap', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            quoteResponse,
            userPublicKey,
            wrapAndUnwrapSol: true,
            computeUnitPriceMicroLamports: 'auto'
        })
    });
    const { swapTransaction } = await response.json();
    if (!swapTransaction) {
        throw new Error('Failed to get swap transaction');
    }
    // Deserialize the transaction
    const swapTransactionBuf = Uint8Array.from(atob(swapTransaction)
        .split('')
        .map(c => c.charCodeAt(0)));
    return web3_js_1.VersionedTransaction.deserialize(swapTransactionBuf);
}
