// @lit-protocol/aw-tool-jupiter-swap/dist/src/lib/tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JupiterSwap = void 0;
const zod_1 = require("zod");
const aw_tool_1 = require("@lit-protocol/aw-tool");
const policy_1 = require("./policy");
const ipfs_1 = require("./ipfs");
/**
 * Zod schema for validating `JupiterSwapLitActionParameters`.
 */
const JupiterSwapLitActionSchema = zod_1.z.object({
    pkpEthAddress: zod_1.z
        .string()
        .regex(/^0x[a-fA-F0-9]{40}$/, 'Must be a valid Ethereum address (0x followed by 40 hexadecimal characters)'),
    tokenIn: zod_1.z.string(),
    tokenOut: zod_1.z.string(),
    amountIn: zod_1.z.string(),
});
/**
 * Descriptions of each parameter for the JupiterSwap Lit Action.
 * These descriptions are designed to be consumed by LLMs (Language Learning Models) to understand the required parameters.
 */
const JupiterSwapLitActionParameterDescriptions = {
    pkpEthAddress: 'The Ethereum address of the PKP that will be used to perform the action.',
    tokenIn: 'The Solana contract address of the SPL token you want to send. Must be a valid SPL address.',
    tokenOut: 'The Solana contract address of the SPL token you want to receive. Must be a valid SPL address.',
    amountIn: 'The amount of tokens to send, specified as a string. This should be a decimal number (e.g. "1.5" or "100"). The amount will be automatically adjusted based on the token\'s decimals.',
};
/**
 * Validates the parameters for the JupiterSwap Lit Action.
 * @param params - The parameters to validate.
 * @returns `true` if the parameters are valid, or an array of errors if invalid.
 */
const validateJupiterSwapParameters = (params) => {
    const result = JupiterSwapLitActionSchema.safeParse(params);
    if (result.success) {
        return true;
    }
    // Map validation errors to a more user-friendly format
    return result.error.issues.map((issue) => ({
        param: issue.path[0],
        error: issue.message,
    }));
};
/**
 * Creates a network-specific JupiterSwap tool.
 * @param network - The supported Lit network (e.g., `datil-dev`, `datil-test`, `datil`).
 * @param config - The network configuration.
 * @returns A configured `AwTool` instance for the JupiterSwap Lit Action.
 */
const createNetworkTool = (network, config) => ({
    name: 'JupiterSwap',
    description: `JupiterSwap Tool`,
    ipfsCid: ipfs_1.IPFS_CIDS[network].tool,
    defaultPolicyIpfsCid: ipfs_1.IPFS_CIDS[network].defaultPolicy,
    chain: 'solana',
    parameters: {
        type: {},
        schema: JupiterSwapLitActionSchema,
        descriptions: JupiterSwapLitActionParameterDescriptions,
        validate: validateJupiterSwapParameters,
    },
    policy: policy_1.JupiterSwapPolicy,
});
/**
 * Exports network-specific JupiterSwap tools.
 * Each tool is configured for a specific Lit network (e.g., `datil-dev`, `datil-test`, `datil`).
 */
exports.JupiterSwap = Object.entries(aw_tool_1.NETWORK_CONFIGS).reduce((acc, [network, config]) => ({
    ...acc,
    [network]: createNetworkTool(network, config),
}), {});
