// @virtuals-protocol/game/dist/plugins/alloraPlugin/src/example.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
const allora_sdk_1 = require("@alloralabs/allora-sdk");
const game_1 = require("@virtuals-protocol/game");
const _1 = __importDefault(require("."));
const alloraPlugin = new _1.default({
    apiClientConfig: {
        chainSlug: process.env.ALLORA_CHAIN_SLUG,
        apiKey: process.env.ALLORA_API_KEY, // Default key: UP-17f415babba7482cb4b446a1
    },
});
// Create an agent with the worker
const agent = new game_1.GameAgent((_a = process.env.GAME_API_KEY) !== null && _a !== void 0 ? _a : "", {
    name: "Allora Worker",
    goal: "Get the 5m price inference for BTC and Luna from Allora Network.",
    description: `You are an AI agent specialized in Allora Network.
You are able to get price inferences from Allora Network and provide users insights into future price of different crypto assets.
You are able to get details about the topics deployed on Allora Network and provide users insights into the topics.
For all the active topics, you are able to get the latest inference using the topic id.
The available assets for price inferences worker are ${Object.values(allora_sdk_1.PriceInferenceToken).join(", ")};
for the following timeframes: ${Object.values(allora_sdk_1.PriceInferenceTimeframe).join(", ")}.
If a price inference is not available for a specific asset and timeframe,
you should determine the topic id for the asset and timeframe and use the topics inferences worker to get the latest inference 
for the specified asset and timeframe. This will return the equivalent of a price inference for the asset and timeframe.
  `,
    workers: [
        alloraPlugin.getWorker({}),
    ],
});
(() => __awaiter(void 0, void 0, void 0, function* () {
    agent.setLogger((agent, message) => {
        console.log(`-----[${agent.name}]-----`);
        console.log(message);
        console.log("\n");
    });
    yield agent.init();
    while (true) {
        yield agent.step({
            verbose: true,
        });
    }
}))();
