// @virtuals-protocol/game/dist/plugins/ensoPlugin/src/example.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const game_1 = require("@virtuals-protocol/game");
const _1 = require(".");
const chains_1 = require("viem/chains");
const viem_1 = require("viem");
const accounts_1 = require("viem/accounts");
const dotenv = __importStar(require("dotenv"));
dotenv.config();
const account = (0, accounts_1.privateKeyToAccount)(process.env.WALLET_PRIVATE_KEY);
const publicClient = (0, viem_1.createPublicClient)({
    transport: (0, viem_1.http)(process.env.RPC_PROVIDER_URL),
    chain: chains_1.base,
});
const walletClient = (0, viem_1.createWalletClient)({
    account: account,
    transport: (0, viem_1.http)(process.env.RPC_PROVIDER_URL),
    chain: chains_1.base,
});
(() => __awaiter(void 0, void 0, void 0, function* () {
    var _a;
    const ensoActionsWorker = yield (0, _1.getEnsoWorker)({
        wallet: walletClient,
        publicClient,
        apiKey: process.env.ENSO_API_KEY || "1e02632d-6feb-4a75-a157-documentation",
    });
    const agent = new game_1.GameAgent((_a = process.env.GAME_API_KEY) !== null && _a !== void 0 ? _a : "", {
        name: "Enso Actions Agent",
        goal: "Find the best route between two tokens and execute it",
        description: "An agent that finds the best route between tokens and executes it",
        workers: [ensoActionsWorker],
    });
    agent.setLogger((agent, message) => {
        console.log(`-----[${agent.name}]-----`);
        console.log(`${message}\n`);
    });
    yield agent.init();
    const agentWorker = agent.getWorkerById(ensoActionsWorker.id);
    const task = `Swap 0.0002 WETH (address 0x4200000000000000000000000000000000000006) for USDC (address 0x833589fcd6edb6e08f4c7c32d4f71b54bda02913)`;
    yield agentWorker.runTask(task, { verbose: true });
}))();
