// @virtuals-protocol/game@0.1.9/dist/plugins/ensoPlugin/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC20_ABI_MIN = exports.ENSO_ETH = exports.ENSO_SUPPORTED_CHAINS = void 0;
exports.ENSO_SUPPORTED_CHAINS = new Map([
    [1, "ethereum"],
    [10, "optimism"],
    [56, "bnb chain"],
    [100, "gnosis"],
    [137, "polygon"],
    [324, "zksync"],
    [8453, "base"],
    [42161, "arbitrum"],
    [43114, "avalanche"],
    [59144, "linea"],
]);
exports.ENSO_ETH = "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee";
exports.ERC20_ABI_MIN = [
    {
        constant: false,
        inputs: [
            {
                name: "_spender",
                type: "address",
            },
            {
                name: "_value",
                type: "uint256",
            },
        ],
        name: "approve",
        outputs: [
            {
                name: "",
                type: "bool",
            },
        ],
        payable: false,
        stateMutability: "nonpayable",
        type: "function",
    },
];
