// @lit-protocol/aw-tool-sign-ecdsa@0.1.0-19/dist/src/lib/tool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SignEcdsa = void 0;
const zod_1 = require("zod");
const aw_tool_1 = require("@lit-protocol/aw-tool");
const policy_1 = require("./policy");
const ipfs_1 = require("./ipfs");
/**
 * Zod schema for validating `SignEcdsaLitActionParameters`.
 * Ensures that the message is a valid string.
 */
const SignEcdsaLitActionSchema = zod_1.z.object({
    pkpEthAddress: zod_1.z
        .string()
        .regex(/^0x[a-fA-F0-9]{40}$/, 'Must be a valid Ethereum address (0x followed by 40 hexadecimal characters)'),
    message: zod_1.z.string(),
});
/**
 * Descriptions of each parameter for the Signing ECDSA Lit Action.
 * These descriptions are designed to be consumed by LLMs (Language Learning Models) to understand the required parameters.
 */
const SignEcdsaLitActionParameterDescriptions = {
    pkpEthAddress: 'The Ethereum address of the PKP that will be used to sign the message.',
    message: 'The message you want to sign.',
};
/**
 * Validates the parameters for the Signing ECDSA Lit Action.
 * @param params - The parameters to validate.
 * @returns `true` if the parameters are valid, or an array of errors if invalid.
 */
const validateSignEcdsaParameters = (params) => {
    const result = SignEcdsaLitActionSchema.safeParse(params);
    if (result.success) {
        return true;
    }
    // Map validation errors to a more user-friendly format
    return result.error.issues.map((issue) => ({
        param: issue.path[0],
        error: issue.message,
    }));
};
/**
 * Creates a network-specific SignEcdsa tool.
 * @param network - The supported Lit network (e.g., `datil-dev`, `datil-test`, `datil`).
 * @param config - The network configuration.
 * @returns A configured `AwTool` instance for the Signing ECDSA Lit Action.
 */
const createNetworkTool = (network, config) => ({
    name: 'SignEcdsa',
    description: `A Lit Action that signs a message with an allowlist of message prefixes.`,
    ipfsCid: ipfs_1.IPFS_CIDS[network].tool,
    defaultPolicyIpfsCid: ipfs_1.IPFS_CIDS[network].defaultPolicy,
    parameters: {
        type: {},
        schema: SignEcdsaLitActionSchema,
        descriptions: SignEcdsaLitActionParameterDescriptions,
        validate: validateSignEcdsaParameters,
    },
    policy: policy_1.SignEcdsaPolicy,
});
/**
 * Exports network-specific SignEcdsa tools.
 * Each tool is configured for a specific Lit network (e.g., `datil-dev`, `datil-test`, `datil`).
 */
exports.SignEcdsa = Object.entries(aw_tool_1.NETWORK_CONFIGS).reduce((acc, [network, config]) => ({
    ...acc,
    [network]: createNetworkTool(network, config),
}), {});
