// @lit-protocol/aw-tool-enso@0.1.0-19/dist/src/lib/lit-actions/utils/approve.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createApproveTx = void 0;
/**
 * Generate and return ERC20 Approval tx
 * @param {EnsoClient} ensoClient - Enso API Client
 * @param {number} chainId - Chain ID of blockchain network
 * @param {any} gasData - Gas data of chain
 * @param {string} sender - Address from whom the transaction will be executed
 * @param {string} token - ERC20 Token address
 * @param {string} amount - amount to approve
 * @returns {Promise<{bestQuote: any;bestFee: number;amountOutMin: any;}>} The best quote and fee tier.
 */
const createApproveTx = async (ensoClient, chainId, gasData, sender, token, amount) => {
    const approve = await ensoClient.getApprovalData({
        chainId,
        fromAddress: sender,
        amount,
        tokenAddress: token,
    });
    return {
        to: approve.tx.to,
        data: approve.tx.data,
        value: '0x0',
        gasLimit: ((BigInt(approve.gas) * 15n) / 10n).toString(), // Increase gas by 50%
        maxFeePerGas: gasData.maxFeePerGas,
        maxPriorityFeePerGas: gasData.maxPriorityFeePerGas,
        nonce: gasData.nonce,
        chainId: chainId.toString(),
        type: 2,
    };
};
exports.createApproveTx = createApproveTx;
